### ISSUE TITLE
[FR] Conditional Termination in CustomIntegrator

### ISSUE THREAD

**Issue by kyleabeauchamp** [pandegroup/openmm#941]
In #938, we identified conditional termination as one possible way to extend the ability of CustomIntegrator to handle logic and decision making.  The key idea is that this might be way to create somewhat more complicated integrators, including ones that involve slightly more complicated steps of hamiltonian monte carlo.  

In #939, we discussed the following possible approach, where the user first does saves a global computation step into a variable, then conditionally terminates the current dynamics step, as outlined below:

```
integrator.addComputeGlobal(variable, expression)
integrator.addConditionalTermination(variable)
```

**Comment by jchodera** [comment]
It would likely be simpler for users (but likely not Implementors) to simplify the API for this to
```
integrator.addConditionalTermination(expression)
```

Support for more general conditionals could also be very useful (though does not need to be exclusive with this feature).  In #938, @peastman had suggested a more general conditional framework such as
```
integrator.beginIfBlock("accept", CustomIntegrator.EQUALS, "1") # Only execute the block if accept==1
# Insert more commands here
integrator.endBlock()
```
where there would presumably be a family of comparators such as `EQUALS`, `LESSTHAN`, `GREATERTHAN`, `NOTEQUALS`, etc.  This starts to get a little clunky, making readability a bit difficult, but would still be very useful.

Other kinds of control flow would also be useful, such as `while` loops:
```
integrator.beginWhileBlock("continue", CustomIntegrator.EQUALS, "1") # Only execute the block as long as continue==1
# Insert more commands here
integrator.endBlock()
```

Not quite sure how to avoid clunkiness and improve readability without going to a parser that actually processes some sort of made-up language (e.g. using Bison and Flex).

**Comment by kyleabeauchamp** [comment]
Given that my available time is going to decrease substantially after friday, I vote for the approach that is easiest to design and implement.

**Comment by jchodera** [comment]
Right, that's the other thing:
* There is an urgent need to have at least `addConditionalTermination` added so @kyleabeauchamp can finish a paper before he leaves on Friday.
* More general conditional extensions need more discussion, but don't need to hold up this feature.

**Comment by rmcgibbo** [comment]
A while ago, I tried to write a custom integrator for Bernie Brooks' SGLD,
and I started by writing a little minilanguage with pyparsing that
'compliled' a multiline string of code to custom integrator calls.

If we end up going down the road of adding conditionals (I'm not weighing
in either way), we might not consider extending the syntax a little bit to
be more natural.

**Comment by jchodera** [comment]
> If we end up going down the road of adding conditionals (I'm not weighing in either way), we might not consider extending the syntax a little bit to be more natural.

I think this would be a good direction to go in as we start adding more complex control flow.  This might be an alternative way to construct `CustomIntegrator` objects.

@rmcgibbo: Can you give a pointer to your mini-language, for inspiration?

**Comment by peastman** [comment]
If Kyle needs to finish up a paper by Friday, implement whatever is convenient for your own use!  But that doesn't mean we should merge it into the main repository.  Let's design this properly and take the time to do it right.

> integrator.beginWhileBlock("continue", CustomIntegrator.EQUALS, "1") # Only execute the block as long as continue==1

Yes, that's precisely what I had in mind.

**Comment by kyleabeauchamp** [comment]
So how about we first discuss the barebones modifications required to implement the allegedly simplest idea, e.g. 

```
integrator.addConditionalTermination(expression)
```

I think what we need from @peastman is his intuition for helping to predict any roadblocks on implementing that minimal approach.

**Comment by kyleabeauchamp** [comment]
*edit* I think we've all concluded that the simplest to implement would be a conditional termination on a variable.  I've edited my comment to say as much.

**Comment by peastman** [comment]
That's the minimum change needed to support this one use case.  We still need to consider the full range of use cases, and the full range of possible implementations.

- Should we provide generic "if" blocks?
- Should we provide generic "while" blocks?
- If we have "if" blocks, should we still have a "conditional termination" step, or just a "terminate" step that you can use inside "if" blocks, or no "terminate" step at all?
- If we have "while" blocks, should there be a "break" step?  Should it be conditional?
- Should we consider allowing a condition to be attached to _any_ step?

Once we have "if" and "while", I don't think either the "terminate" or "break" steps is actually required for functionality.  You can implement any logic even without them.  The question then is whether they add enough convenience to justify the extra complexity.  As for adding conditions to individual steps, that is purely a convenience.  At best, it means adding one step (e.g. terminate with condition) instead of three steps (if, terminate, end if).  But if you have multiple steps all based on the same condition, with "terminate" being the last one, you'll need an "if" block anyway.  Having to put a condition on the "terminate" is then just an inconvenience (and possibly an inefficiency).

**Comment by kyleabeauchamp** [comment]
Here is my perspective:

1.  With a small amount of help and discussion, I may be able to implement the minimal approach that I have suggested.  
2.  I'm not asking that we *merge* my implementation--we are free to re-implement a better or more general version later.  
3.  Once I implement the minimal version, I'll have a better idea of how *useful* this feature is.  

So I guess my question is: does anyone have a suggestion on a better "quick and dirty" solution.  If not, does anyone have any suggestions on implementing the "quick and dirty" solution.  Again, I'm not asking to merge this.

**Comment by rmcgibbo** [comment]
@jchodera: here's the minilanguage I was working with to implement SGLD. (Note: I don't think the integrator actually works, still some bugs). https://gist.github.com/rmcgibbo/31525aec58ac3a1ad1e6#file-sgld-py-L55-L63

**Comment by jchodera** [comment]
@kyleabeauchamp: I think the logic you suggest, where you only add
```
integrator.addConditionalTermination(variablename)
```
is probably the simplest way to go for your XCHMC.

Longer-term:
* To answer @peastman's question, what if we start with `if` and `while`?  We may not need `terminate` with these two pieces of control flow.  If things become a mess without `terminate`, `break`, and the other suggestions, then we can add those.  We could even prototype a few integrators just to see.
* @rmcgibbo's minilanguage looks pretty nice!  It seems useful enough and simple enough that this could be easily added to the Python layer.

**Comment by rmcgibbo** [comment]
yeah, I was thinking that if conditional blocks get added (`if`, `while`), then something like that would get even more useful.

**Comment by kyleabeauchamp** [comment]
PS: I may have just thought of a way to implement XCHMC without the conditional termination.  Basically, instead of termination I can just throw away any extra dynamics that I do.  If I minimize the amount of CPU time spent doing the discarded sampling, I'll be OK in the end.

**Comment by rmcgibbo** [comment]
also, FWIW, pyparsing (the parsing library I used for that code; we also use it in mdtraj for the atom selection dsl) is a single MIT-licensed file, so it's relatively convenient to just plop into the source tree, if we end up going down that route.

**Comment by peastman** [comment]
I've created a draft of the API for this: https://github.com/peastman/openmm/commit/0f21cca739f67315b8f83ee12c8b1cbf8239d531.  Does that look like a good design?  It's going to be a lot of work to implement, so I want to make sure the API is right before putting lots of work into it.

**Comment by swails** [comment]
LGTM

**Comment by jchodera** [comment]
Looks nice, @peastman!  Perhaps add an example of a `while` block?  The example could simply be a Langevin integrator that keeps trying until the coordinates are not `nan`.  (Or would that require an `isnan(x)` function?)

**Comment by swails** [comment]
`isnan(x)` should be as simple as `if (x != x)`...

**Comment by jchodera** [comment]
I think it was more difficult than that (see #895).  The recommendation was `abs(step(x)-step(1-x))`.

**Comment by swails** [comment]
But that was before @peastman added the `ifBlock`.  That weird gymnastics shouldn't be necessary anymore, right?

The only way this doesn't work is if Lepton optimizes it out by seeing something it assumes will always be false.

**Comment by peastman** [comment]
That ought to work.  But I'm not going to add an example of checking for nan.  That's an obscure and advanced thing that should only be done by people who really know what they're doing!  I don't want to encourage it, especially among inexperienced users.

**Comment by swails** [comment]
> That ought to work.

"That" referring to... ?

**Comment by peastman** [comment]
x != x

**Comment by jchodera** [comment]
>  That's an obscure and advanced thing that should only be done by people who really know what they're doing! I don't want to encourage it, especially among inexperienced users.

That sounds like a great argument for an `isnan(x)` feature request.  :)

**Comment by peastman** [comment]
I didn't mean that particular method of checking for nans.  I mean that in most cases, you should make sure that nothing ever becomes nan, and if anything does, it's too late to do anything except exit the simulation.  There are exceptions, like Kyle's integrator, but that's for very advanced users.  Most people shouldn't be trying to implement anything like that.

**Comment by swails** [comment]
He was saying it's an argument *against* it.  Only an expert should *expect* to get a NaN in their simulation and intend to deal with it.  Adding an `isnan` may be giving too much rope, and the workaround

```Python
integrator.ifBlock("x != x")
# handle the NaN
integrator.endBlock()
```

is simple enough that it's not a huge barrier to being an expert ;).

**Comment by jchodera** [comment]
Well, 
```
integrator.ifBlock("x != x")
```
would not work because `x` is a per-dof variable, not a global variable.

The idea of checking for `nan` in Langevin integrators is definitely an advanced topic and not something we should encourage, but would be reasonable in a Metropolis Monte Carlo integrator.  But I suppose now that we have `select`, one could instead use that to set the `acceptanceProbability` to zero if the change in energy is `nan`, rather than checking for `nan` directly in the acceptance step.

Perhaps there would be utility in allowing `ifBlock(conditional)` conditionals to operate on per-dof variables, at least indirectly?  Numpy has `any(x)` and `all(x)`, though I suppose these can currently be coded as `computeSum` of `step(x)`, though `all(x)` would have to be something like checking that the sum of `step(x)` is equal to the number of degrees of freedom.

**Comment by swails** [comment]
`x` was a generic variable, not positions in my example.  The energy should be a single number, so for `x` equal to the energy, my example should still work, right?

**Comment by jchodera** [comment]
Yes, sorry for the confusion there.

**Comment by peastman** [comment]
CustomIntegrator basically defines a SIMT architecture.  A lot like programming a GPU, in fact!  All threads (i.e. degrees of freedom) execute the same sequence of computations, but the input and output values for each thread are different.  If you need different threads to do different calculations, you have to use predication, which is effectively what select() does.  In fact, I chose that name to be consistent with the select() function in OpenCL, which does the same thing.

And yes, you could use computeSum to implement an "any" or "all" operation.

**Comment by peastman** [comment]
I have the reference implementation working and checked in at https://github.com/peastman/openmm/tree/flowcontrol, if you want to try it out.

**Comment by peastman** [comment]
Fixed by #1048.
